tests/test_bgp_af.rb

```diff
@@ -959,4 +959,89 @@ def test_utils_delta_add_remove
     assert_equal(expected, result, 'Test 5. delta mismatch')
   end
+
+  # test route_target
+  def test_route_target
+    afs = [%w(ipv4 unicast), %w(ipv6 unicast)]
+    afs.each do |af|
+      route_target(55, 'red', af)
+    end
+  end
+
+  def route_target(asn, vrf, af)
+    # Common test for route-target providers. Tests evpn and non-evpn.
+
+    bgp_af = RouterBgpAF.new(asn, vrf, af)
+
+    # test route target both auto and route target both auto evpn
+    refute(bgp_af.default_route_target_both_auto,
+           'default value for route target both auto should be false')
+
+    refute(bgp_af.default_route_target_both_auto_evpn,
+           'default value for route target both auto evpn should be false')
+
+    bgp_af.route_target_both_auto = true
+    assert(bgp_af.route_target_both_auto, "vrf context #{vrf} af #{af}: "\
+           'bgp_af route-target both auto should be enabled')
+
+    bgp_af.route_target_both_auto = false
+    refute(bgp_af.route_target_both_auto, "vrf context #{vrf} af #{af}: "\
+           'bgp_af route-target both auto should be disabled')
+
+    bgp_af.route_target_both_auto_evpn = true
+    assert(bgp_af.route_target_both_auto_evpn, "vrf context #{vrf} af #{af}: "\
+           'bgp_af route-target both auto evpn should be enabled')
+
+    bgp_af.route_target_both_auto_evpn = false
+    refute(bgp_af.route_target_both_auto_evpn, "vrf context #{vrf} af #{af}: "\
+           'bgp_af route-target both auto evpn should be disabled')
+
+    opts = [:import, :export]
+
+    # Master list of communities to test against
+    master = '1:1 2:2 3:3 4:5'
+
+    # Test 1: both/import/export when no commands are present. Each target
+    # option will be tested with and without evpn (6 separate types)
+    should = master.clone
+    route_target_tester(bgp_af, af, opts, should, 'Test 1')
+
+    # Test 2: remove half of the entries
+    should = '1:1 4:4'
+    route_target_tester(bgp_af, af, opts, should, 'Test 2')
+
+    # Test 3: restore the removed entries
+    should = master.clone
+    route_target_tester(bgp_af, af, opts, should, 'Test 3')
+
+    # Test 4: 'default'
+    should = bgp_af.default_route_target_import
+    route_target_tester(bgp_af, af, opts, should, 'Test 4')
+  end
+
+  def route_target_tester(bgp_af, af, opts, should, test_id)
+    # First configure all four property types
+    opts.each do |opt|
+      # non-evpn
+      bgp_af.send("route_target_#{opt}=", should)
+    end
+
+    # Now check the results
+    opts.each do |opt|
+      # non-evpn
+      result = bgp_af.send("route_target_#{opt}")
+      assert_equal(should, result,
+                   "#{test_id} : #{af} : route_target_#{opt}")
+    end
+
+    opts.each do |opt|
+      bgp_af.send("route_target_#{opt}_evpn=", should)
+    end
+
+    opts.each do |opt|
+      result = bgp_af.send("route_target_#{opt}_evpn")
+      assert_equal(should, result,
+                   "#{test_id} : #{af} : route_target_#{opt}_evpn")
+    end
+  end
 end
```

Comment:
We'll need some additional tests for the delta_add_remove method since it's been modified to handle depth == 1.

Take a look at test_utils_delta_add_remove and replicate it to test depth 1, e.g.:

```
def   def test_utils_delta_add_remove_depth_1
  ...
    should = ['1:1', '2:2', '3:3', '4:4', '5:5', '6:6']

   current = []
    expected = { add: should, remove: [] }
    result = Utils.delta_add_remove(should, current)
    assert_equal(expected, result, 'Test 1. delta mismatch')
  ...
  ... more tests
```